src/types-and-traits.rst

```diff
@@ -1995,6 +1995,39 @@ obtained as follows:
    #. :dp:`fls_ju4ypa5ysga0`
       Continue with the next source :t:`type`.
 
+
+.. _fls_emcNJzl2tHSA:
+
+Structural Equality
+~~~~~~~~~~~~~~~~~~~
+
+.. rubric:: Legality Rules
+
+:dp:`fls_uVTpA7gbLCYX`
+A :t:`structurally equal` :t:`type` can be :t:`pattern matched <pattern
+matching>` on via the use of an :t:`associated constant` or :t:`constant`.
+
+:dp:`fls_emcNJzl2tHSA`
+:t:`[Bool]`, :t:`Char`, :t:`[integer type]s`, :t:`str`, :t:`[function pointer
+type]s` and :t:`[raw pointer type]s` are :t:`structurally equal`.
+
+:dp:`fls_HpWSAfaTA1Dz`
+An :t:`abstract data type` is :t:`structurally equal` if it implements the
+:t:`[trait]s` :std:`core::marker::PartialEq` and `core::marker::Eq` via the use
+of the :t:`derive` attribute.
+
+:dp:`fls_5RcnETrW6f9m`
+:t:`[Slice type]s` and :t:`[array type]s` are :t:`structurally equal` if the
+:t:[`element type]` contained within is :t:`structurally equal`.
```

Comment:
I would rephrase this as

An array type or a spice type is structurally equal if its element type is structurally equal.